internal/jsonopts/options.go

```diff
@@ -170,7 +169,13 @@ func (dst *Struct) Join(srcs ...Options) {
 				dst.FormatDepth = src.FormatDepth
 			}
 		default:
-			JoinUnknownOption(dst, src)
+			unknown := true
+			if joiner, ok := src.(interface{ Join(*Struct, Options) bool }); ok {
+				unknown = joiner.Join(dst, src)
+			}
+			if unknown {
+				JoinUnknownOption(dst, src)
+			}
```

Comment:
This is the core logic of the PR. Everything else is making private symbols public that are needed to use this logic.